### DISCUSSION TITLE
API | Having syntax issues when adding tags on records (v3.3.8)

### DISCUSSION THREAD

**Discussion by eskdera** [netbox-community/netbox#10978] [Q&A]
Hi, we have recently upgraded our netbox to v3.3.8 and the previous syntax for adding tags to a (tenant) record through a post request doesn't work anymore. this used to work before (per https://demo.netbox.dev/api/docs/):
`                    var body = JSON.stringify({
                        'name': 'name',
                        'slug': 'slug',
                        'tags': [
                            { "name": "slug",
                              "slug": "slug"}] });`
I've tried other syntaxes as well but had no luck adding a tag while adding a tenant record through the API.
I get this error:
Cannot add ['tag'] (<class 'list'>). Expected <class 'django.db.models.base.ModelBase'> or str.\n\nPython version: 3.9.6\nNetBox version: 3.3.8<\/pre>

I appreciate the help.

**Comment by candlerb** [comment]
> we have recently upgraded our netbox to v3.3.8

From which previous version?

**Reply by eskdera** [reply]
2.10.1

**Reply by candlerb** [reply]
The API does change between versions. The following works for me, with v3.3.8:

```
# curl -gsS -H "Authorization: Token $API_TOKEN" "https://$NETBOX/api/tenancy/tenants/1/" | jq '.tags'
[]
# curl -d '{"tags":[{"name":"gateway"}]}' -X PATCH -gsS -H "Content-Type: application/json" -H "Authorization: Token $API_TOKEN" "https://$NETBOX/api/tenancy/tenants/1/"
... snip response ...
# curl -gsS -H "Authorization: Token $API_TOKEN" "https://$NETBOX/api/tenancy/tenants/1/" | jq '.tags'
[
  {
    "id": 9,
    "url": "https://XXXX/api/extras/tags/9/",
    "display": "gateway",
    "name": "gateway",
    "slug": "gateway",
    "color": "009688"
  }
]
# curl -d '{"tags":[]}' -X PATCH -gsS -H "Content-Type: application/json" -H "Authorization: Token $API_TOKEN" "https://$NETBOX/api/tenancy/tenants/1/"
... snip response ...
# curl -gsS -H "Authorization: Token $API_TOKEN" "https://$NETBOX/api/tenancy/tenants/1/" | jq '.tags'
[]
# 
```

This suggests the problem is in your code somewhere. Inspect the request body you are sending; check you are using the right method (e.g. PATCH); check you are using the right URL for the tenant.

EDIT: note that a POST to `/tenancy/tenants/` is used to [create](https://docs.netbox.dev/en/stable/integrations/rest-api/#creating-a-new-object) a new tenant. You will need a PUT or PATCH to `/tenancy/tenants/NNN/` to [update](https://docs.netbox.dev/en/stable/integrations/rest-api/#updating-an-object) an existing tenant.

**Reply by eskdera** [reply]
@candlerb The issue must be with the way data is being passed to tags because when I remove `"tags": [ {"name": "test","slug": "test" }]` from the request it works with no errors (I tested the request in postman as well same deal) with tags in the request I get the following exception.
The complete exception is provided below:		
<pre class="block"><strong>&lt;class &#x27;ValueError&#x27;&gt;</stron
g><br />
Cannot add [&#x27;test&#x27;] (&lt;class &#x27;list&#x27;&gt;). Expected &lt;class &#x27;django.db.models.base.ModelBase&#x27;&gt; or str.

Python version: 3.9.6
NetBox version: 3.3.8</pre>
And I'm using POST since I want to create a new tenant.

**Reply by candlerb** [reply]
What we need to see is the *entire actual PATCH body*.  My guess is that you're not generating it correctly from PowerShell.

Please see if you can print the body before sending it.  And if you can't do that, then use tcpdump to capture it at the Netbox end:

```
tcpdump -i eth0 -nn -s0 -A tcp port 8001
```

**Comment by mtinberg** [comment]
If you are setting or removing related values (eg. tags) isn't that just an array of id number or name strings and not a hash with all the field values like is returned

eg. "tags": [1, 7]
or  "tags": ["foo", "bar"]
not "tags": [{"name": "foo", "id": 1},{"name": "bar", "id": 7}]


—

Mark Tinberg ***@***.***>

Division of Information Technology-Network Services

University of Wisconsin-Madison


________________________________

**Reply by eskdera** [reply]
@mtinberg Back at v 2.10.1 it used to be like "tags": [{"name": "foo", "id": 1},{"name": "bar", "id": 7}] no errors. 
Just now I tried "tags": [1] and I still get the same error.

**Reply by candlerb** [reply]
I have tested these and they all work:

```
{"tags":[8]}
{"tags":[{"slug":"gateway"}]}
{"tags":[{"name":"gateway", "slug":"gateway"}]}
```

I don't know about `[{"id":8}]`, but I don't see why that wouldn't work.
